include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp

```diff
@@ -332,107 +367,142 @@ namespace nil {
+                std::vector<constraint_type> nine_block;
+                std::vector<constraint_type> bit_line;
+                constraint_type first_add_W1 = var(component.W(1),0) + 1;
+                constraint_type first_add_W2 = var(component.W(2),0) * (1-var(component.W(2),0));
+                constraint_type first_W4 = var(component.W(3),0)*var(component.W(7),0) + 1-var(component.W(7),0) - var(component.W(4),0);
 
                 // Store typical constraints for every column
-                six_block.resize(WA);
+                nine_block.resize(WA);
                 for(std::size_t j = 0; j < l; j++) {
-                    var W0 = var(component.W(6*j),0),
-                        W1 = var(component.W(6*j + 1),0),
-                        W2 = var(component.W(6*j + 2),0),
-                        W3 = var(component.W(6*j + 3),0),
-                        W4 = var(component.W(6*j + 4),0),
-                        W5 = var(component.W(6*j + 5),0),
-                        W0next = var(component.W(6*((j+1) % l)),(j+1)/l),
-                        W1prev = var(component.W(6*((l+j-1) % l) + 1), -(j == 0)),
-                        W2prev = var(component.W(6*((l+j-1) % l) + 2), -(j == 0)),
-                        W3next = var(component.W(6*((j+1) % l) + 3),(j+1)/l);
-
-                    six_block[6*j]   = W0 - 2*W0next - W5;
-                    six_block[6*j+1] = W1 - W1prev * W1prev;
-                    six_block[6*j+2] = W2 - W2prev * (W1*W5 + 1-W5);
-                    six_block[6*j+3] = W3 - W3next * W3next;
-                    six_block[6*j+4] = W3 + W4;
-                    six_block[6*j+5] = (1 - W5) * W5;
+                    var W0 = var(component.W(9*j),0),
+                        W1 = var(component.W(9*j + 1),0),
+                        W2 = var(component.W(9*j + 2),0),
+                        W3 = var(component.W(9*j + 3),0),
+                        W4 = var(component.W(9*j + 4),0),
+                        W5 = var(component.W(9*j + 5),0),
+                        W6 = var(component.W(9*j + 6),0),
+                        W7 = var(component.W(9*j + 7),0),
+                        W8 = var(component.W(9*j + 8),0),
+                        W0next = var(component.W(9*((j+1) % l)),(j+1)/l),
+                        W1next = var(component.W(9*((j+1) % l) + 1),(j+1)/l),
+                        W2next = var(component.W(9*((j+1) % l) + 2),(j+1)/l),
+                        W3prev = var(component.W(9*((l+j-1) % l) + 3), -(j == 0)),
+                        W4prev = var(component.W(9*((l+j-1) % l) + 4), -(j == 0)),
+                        W5next = var(component.W(9*((j+1) % l) + 5),(j+1)/l);
+
+                    nine_block[9*j]   = W0 - 2*W0next - W7;
+                    nine_block[9*j+1] = W1 - 2*W1next - W8;
+                    nine_block[9*j+2] = (W8 - W7)*(1 - W2next)*(1 + W2next) + (W2next * W2next * W2next) - W2;
```

Comment:
Why is the cubing required here? It seems to me that just doing `+ W2next` should suffice.

Reply:
That's right. No idea, why I wanted to cube so much.